From e96b79c9e4cd6c40ba6e5fe495904f07a95fb909 Mon Sep 17 00:00:00 2001
From: Thomas Deutschmann <whissi@gentoo.org>
Date: Wed, 13 Dec 2017 01:02:12 +0100
Subject: [PATCH 49/49] loff_t and __WORDSIZE includes for MUSL

Rewrite of Justin Keogh's patch [Link 1] to fix build problems
on ARM.

Link 1: https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=5f5abc0f1b036921d6eb5b0f434c960ed280619f
Fixes: https://bugs.gentoo.org/616364
---
 configure.in  | 2 +-
 ddr_ctrl.h    | 3 +++
 ffs.h         | 3 +++
 fiemap.h      | 4 ++++
 fmt_no.h      | 2 ++
 fstrim.h      | 3 +++
 libddr_hash.c | 4 ++++
 libddr_lzo.c  | 3 +++
 libddr_null.c | 3 +++
 sha512.h      | 4 ++++
 10 files changed, 30 insertions(+), 1 deletion(-)

diff --git a/configure.in b/configure.in
index eb2e813..355ea84 100644
--- a/configure.in
+++ b/configure.in
@@ -6,7 +6,7 @@ AC_C_INLINE
 AC_HEADER_STDC
 #AC_PROG_INSTALL
 #CFLAGS="$CFLAGS -DHAVE_CONFIG_H"
-AC_CHECK_HEADERS([fallocate.h dlfcn.h unistd.h attr/xattr.h sys/acl.h sys/ioctl.h endian.h linux/fs.h linux/fiemap.h stdint.h lzo/lzo1x.h openssl/evp.h linux/random.h malloc.h sched.h sys/statvfs.h])
+AC_CHECK_HEADERS([fallocate.h dlfcn.h unistd.h attr/xattr.h sys/acl.h sys/ioctl.h sys/reg.h endian.h linux/fs.h linux/fiemap.h stdint.h lzo/lzo1x.h openssl/evp.h linux/random.h malloc.h sched.h sys/statvfs.h])
 AC_CHECK_FUNCS([ffs ffsl basename fallocate64 splice getopt_long open64 pread pread64 lseek64 stat64 posix_fadvise posix_fadvise64 __builtin_prefetch htobe64 feof_unlocked getline getentropy getrandom posix_memalign valloc sched_yield fstatvfs __builtin_cpu_supports])
 AC_CHECK_LIB(dl,dlsym)
 AC_CHECK_LIB(fallocate,linux_fallocate64)
diff --git a/ddr_ctrl.h b/ddr_ctrl.h
index ac71e4f..58cffd5 100644
--- a/ddr_ctrl.h
+++ b/ddr_ctrl.h
@@ -7,6 +7,9 @@
  *  License: GNU GPLv2 or v3
  */
 
+#define _GNU_SOURCE
+#include <fcntl.h>
+
 #ifndef _DDR_CTRL_H
 #define _DDR_CTRL_H
 
diff --git a/ffs.h b/ffs.h
index 2215080..c1f3444 100644
--- a/ffs.h
+++ b/ffs.h
@@ -28,6 +28,9 @@
 #include <endian.h>
 #endif
 
+#ifdef HAVE_SYS_REG_H
+#include <sys/reg.h>
+#endif
 
 #ifdef HAVE_FFS
 # define myffs(x) ffs(x)
diff --git a/fiemap.h b/fiemap.h
index df1ba95..31cde3b 100644
--- a/fiemap.h
+++ b/fiemap.h
@@ -29,5 +29,9 @@
 
 #endif	/* HAVE_LINUX_FS_H */
 
+#ifdef HAVE_SYS_REG_H
+#include <sys/reg.h>
+#endif
+
 #endif	/* _FIEMAPH */
 
diff --git a/fmt_no.h b/fmt_no.h
index cb0a81b..52ec236 100644
--- a/fmt_no.h
+++ b/fmt_no.h
@@ -1,4 +1,6 @@
 /** Decl for int to str conversion with highlighting */
+#define _GNU_SOURCE
+#include <fcntl.h>
 
 #ifndef _FMT_NO_H
 #define _FMT_NO_H
diff --git a/fstrim.h b/fstrim.h
index 7447061..b9cdcbb 100644
--- a/fstrim.h
+++ b/fstrim.h
@@ -1,3 +1,6 @@
+#define _GNU_SOURCE
+#include <fcntl.h>
+
 #ifndef _FSTRIM_H
 #define _FSTRIM_H
 
diff --git a/libddr_hash.c b/libddr_hash.c
index 7c8cdf8..cdbbd4d 100644
--- a/libddr_hash.c
+++ b/libddr_hash.c
@@ -33,6 +33,10 @@
 #include <unistd.h>
 #include <fcntl.h>
 
+#ifdef HAVE_SYS_REG_H
+#include <sys/reg.h>
+#endif
+
 #include <netinet/in.h>	/* For ntohl/htonl */
 #include <endian.h>
 
diff --git a/libddr_lzo.c b/libddr_lzo.c
index 8d3983e..8c83c04 100644
--- a/libddr_lzo.c
+++ b/libddr_lzo.c
@@ -26,6 +26,9 @@
 #include <errno.h>
 #include <netinet/in.h>
 #include <sys/stat.h>
+#ifdef HAVE_SYS_REG_H
+#include <sys/reg.h>
+#endif
 #include <signal.h>
 #include <lzo/lzo1x.h>
 #include <lzo/lzo1y.h>
diff --git a/libddr_null.c b/libddr_null.c
index 3f0f194..c379961 100644
--- a/libddr_null.c
+++ b/libddr_null.c
@@ -10,6 +10,9 @@
 #include "ddr_ctrl.h"
 #include <string.h>
 #include <stdlib.h>
+#ifdef HAVE_SYS_REG_H
+#include <sys/reg.h>
+#endif
 
 /* fwd decl */
 extern ddr_plugin_t ddr_plug;
diff --git a/sha512.h b/sha512.h
index 4d08043..f54d371 100644
--- a/sha512.h
+++ b/sha512.h
@@ -3,6 +3,10 @@
 
 #include "hash.h"
 
+#ifdef HAVE_SYS_REG_H
+#include <sys/reg.h>
+#endif
+
 void sha512_init(hash_t *ctx);
 void sha384_init(hash_t *ctx);
 void sha512_128(const uint8_t* msg, hash_t* ctx);
-- 
2.15.0

